### ISSUE TITLE
Nodes rolling update when there are changes in NodeTemplate and/or Provisioner config

### ISSUE THREAD

**Issue by marlovil** [aws/karpenter#3895]
### Version

__Karpenter Version:__ v0.27.3

__Kubernetes Version:__ v1.23.16-eks-48e63af

### Expected Behavior

When a change is performed in Node Temlplate and/or Provisioner config and we apply this change, all the nodes launched by Karpenter should be updated performing a node rolling update. Something like this:

1. Launch new nodes (same number of nodes)
2. Cordon old nodes
3. Drain old nodes taking into account apps inside the nodes (specially if there are configured pdbs)
4. Once old nodes are empty, delete them

This is the behavior we have with EKS nodegroups when we modify a launch template (ie. change instance type, disk size, subnet, etc.)

### Actual Behavior

When a change is performed in Node Temlplate and/or Provisioner config and we apply this change, all the nodes launched by Karpenter  are updated, but **at the same time**. All the old nodes are deleted at the same time causing an outage, even if there are applications running in those nodes with a configured PDB. Karpenter doesn't perform a node rolling update. It deletes all old nodes simultaneously.

### Steps to Reproduce the Problem

Modifiy existing Node templates and/or Provisioners with new changes. i.e: change instance type/family, increase disk size,...

### Resource Specs and Logs

n/a

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by marlovil** [comment]
Thanks for your quick answer. 

I will try to show you all the process:

1. Change: `karpenter.k8s.aws/instance-category` from `m` to `c|m`

2. Apply helm chart update
```
helm_release.karpenter: Modifying... [id=karpenter]
helm_release.karpenter: Still modifying... [id=karpenter, 10s elapsed]
helm_release.karpenter: Still modifying... [id=karpenter, 20s elapsed]
helm_release.karpenter: Still modifying... [id=karpenter, 30s elapsed]
helm_release.karpenter: Still modifying... [id=karpenter, 40s elapsed]
helm_release.karpenter: Still modifying... [id=karpenter, 50s elapsed]
helm_release.karpenter: Still modifying... [id=karpenter, 1m0s elapsed]
helm_release.karpenter: Still modifying... [id=karpenter, 1m10s elapsed]
helm_release.karpenter: Modifications complete after 1m18s [id=karpenter]

```

Karpenter logs:

```
2023-05-17T09:19:31.361Z	INFO	controller.termination	cordoned node	{"commit": "d7e22b1-dirty", "node": "ip-10-11-38-111.eu-west-1.compute.internal"}
2023-05-17T09:19:31.371Z	INFO	controller.termination	cordoned node	{"commit": "d7e22b1-dirty", "node": "ip-10-11-38-234.eu-west-1.compute.internal"}
2023-05-17T09:19:31.414Z	INFO	controller.termination	cordoned node	{"commit": "d7e22b1-dirty", "node": "ip-10-11-39-4.eu-west-1.compute.internal"}
2023-05-17T09:19:31.460Z	INFO	controller.termination	cordoned node	{"commit": "d7e22b1-dirty", "node": "ip-10-11-39-44.eu-west-1.compute.internal"}
2023-05-17T09:19:31.872Z	INFO	controller.termination	deleted node	{"commit": "d7e22b1-dirty", "node": "ip-10-11-39-44.eu-west-1.compute.internal"}
2023-05-17T09:19:31.882Z	INFO	controller.termination	deleted node	{"commit": "d7e22b1-dirty", "node": "ip-10-11-38-111.eu-west-1.compute.internal"}
2023-05-17T09:19:31.883Z	INFO	controller.termination	deleted node	{"commit": "d7e22b1-dirty", "node": "ip-10-11-39-4.eu-west-1.compute.internal"}
2023-05-17T09:19:31.885Z	INFO	controller.termination	deleted node	{"commit": "d7e22b1-dirty", "node": "ip-10-11-38-234.eu-west-1.compute.internal"}
2023-05-17T09:20:17.335Z	DEBUG	controller.provisioner	301 out of 622 instance types were excluded because they would breach provisioner limits	{"commit": "d7e22b1-dirty"}
2023-05-17T09:20:17.340Z	DEBUG	controller.provisioner	531 out of 622 instance types were excluded because they would breach provisioner limits	{"commit": "d7e22b1-dirty"}
2023-05-17T09:20:17.346Z	DEBUG	controller.provisioner	613 out of 622 instance types were excluded because they would breach provisioner limits	{"commit": "d7e22b1-dirty"}
2023-05-17T09:20:17.346Z	ERROR	controller.provisioner	Could not schedule pod, incompatible with provisioner "default", no instance type satisfied resources {"cpu":"1","memory":"1Gi","pods":"1"} and requirements karpenter.k8s.aws/instance-category In [c m], karpenter.k8s.aws/instance-generation Exists >4, kubernetes.io/arch In [amd64], nodegroup In [karpenter], karpenter.sh/capacity-type In [on-demand], kubernetes.io/os In [linux], karpenter.k8s.aws/instance-hypervisor In [nitro], karpenter.sh/provisioner-name In [default], topology.kubernetes.io/zone In [eu-west-1b]	{"commit": "d7e22b1-dirty", "pod": "karpenter/debug-deployment-67f7856c4f-qxc5g"}
2023-05-17T09:20:17.346Z	INFO	controller.provisioner	found provisionable pod(s)	{"commit": "d7e22b1-dirty", "pods": 8}
2023-05-17T09:20:17.346Z	INFO	controller.provisioner	computed new machine(s) to fit pod(s)	{"commit": "d7e22b1-dirty", "machines": 2, "pods": 7}
2023-05-17T09:20:17.346Z	INFO	controller.provisioner	launching machine with 3 pods requesting {"cpu":"3205m","memory":"3320Mi","pods":"7"} from types c6in.xlarge, c5.xlarge, c5a.xlarge, c5ad.xlarge, c6a.xlarge and 4 other(s)	{"commit": "d7e22b1-dirty", "provisioner": "default"}
2023-05-17T09:20:17.353Z	INFO	controller.provisioner	launching machine with 4 pods requesting {"cpu":"4205m","memory":"4344Mi","pods":"8"} from types c5n.2xlarge, m6a.2xlarge, m5zn.2xlarge, m5a.2xlarge, m5a.4xlarge and 51 other(s)	{"commit": "d7e22b1-dirty", "provisioner": "default"}
2023-05-17T09:20:19.240Z	INFO	controller.provisioner.cloudprovider	launched instance	{"commit": "d7e22b1-dirty", "provisioner": "default", "id": "i-0f73fa42d71c0d325", "hostname": "ip-10-11-39-169.eu-west-1.compute.internal", "instance-type": "c6a.xlarge", "zone": "eu-west-1b", "capacity-type": "on-demand", "capacity": {"cpu":"4","ephemeral-storage":"15Gi","memory":"7577Mi","pods":"58"}}
2023-05-17T09:20:19.834Z	INFO	controller.provisioner.cloudprovider	launched instance	{"commit": "d7e22b1-dirty", "provisioner": "default", "id": "i-07f4c0061690df9b9", "hostname": "ip-10-11-38-99.eu-west-1.compute.internal", "instance-type": "c6a.2xlarge", "zone": "eu-west-1a", "capacity-type": "on-demand", "capacity": {"cpu":"8","ephemeral-storage":"15Gi","memory":"15155Mi","pods":"58"}}
2023-05-17T09:20:27.336Z	DEBUG	controller.provisioner	319 out of 622 instance types were excluded because they would breach provisioner limits	{"commit": "d7e22b1-dirty"}
2023-05-17T09:20:27.341Z	INFO	controller.provisioner	found provisionable pod(s)	{"commit": "d7e22b1-dirty", "pods": 8}
2023-05-17T09:20:27.341Z	INFO	controller.provisioner	computed new machine(s) to fit pod(s)	{"commit": "d7e22b1-dirty", "machines": 1, "pods": 1}
2023-05-17T09:20:27.341Z	INFO	controller.provisioner	computed 2 unready node(s) will fit 7 pod(s)	{"commit": "d7e22b1-dirty"}
2023-05-17T09:20:27.341Z	INFO	controller.provisioner	launching machine with 1 pods requesting {"cpu":"1205m","memory":"1272Mi","pods":"5"} from types m6i.large, m6idn.large, c5n.2xlarge, c6i.large, m6a.2xlarge and 87 other(s)	{"commit": "d7e22b1-dirty", "provisioner": "default"}
2023-05-17T09:20:29.120Z	INFO	controller.provisioner.cloudprovider	launched instance	{"commit": "d7e22b1-dirty", "provisioner": "default", "id": "i-0c58991e5e9f1af65", "hostname": "ip-10-11-39-145.eu-west-1.compute.internal", "instance-type": "c6a.large", "zone": "eu-west-1b", "capacity-type": "on-demand", "capacity": {"cpu":"2","ephemeral-storage":"15Gi","memory":"3788Mi","pods":"29"}}

```

This is the pdb config for the debug app running in those nodes:
```
$ k describe pdb debug-pdb
Name:             debug-pdb
Namespace:        karpenter
Max unavailable:  1
Selector:         app=debug
Status:
    Allowed disruptions:  1
    Current:              8
    Desired:              7
    Total:                8
Events:                   <none>

``` 
I've attached screen shots during the process:

Initial state:
![1-initial_state](https://github.com/aws/karpenter/assets/45008389/a735ff8a-2c4a-488e-aef1-ccb2ab1133a6)
Nodes deleted:
![2-nodes_deleted](https://github.com/aws/karpenter/assets/45008389/5f02aa5c-307e-4c27-b6f7-2ff148665358)
All pods pending, provisioning new nodes:
![3-provisioning_new_nodes](https://github.com/aws/karpenter/assets/45008389/9807a5ee-c2e6-455e-ae20-8a3aa25a8c7d)
New nodes up:
![4-new-nodes-up](https://github.com/aws/karpenter/assets/45008389/74ac5c37-16b0-4ff4-8b3d-18816bbd12ed)

The helm chart values, with node template and provisioner definitions:
```
replicas: 2

settings:
  aws:
    clusterName: "${cluster_name}"
    defaultInstanceProfile: "${node_instance_profile}"

serviceAccount:
  create: true
  name: "eks-karpenter-oidc-sa"
  annotations:
    eks.amazonaws.com/role-arn: ${iam_role_arn}

nodeSelector:
  kubernetes.io/os: linux
  nodegroup: tools

controller:
  resources:
    requests:
      cpu: 500m
      memory: 512Mi
    limits:
      cpu: 500m
      memory: 512Mi

serviceMonitor:
  enabled: true

extraObjects:
  - apiVersion: karpenter.k8s.aws/v1alpha1
    kind: AWSNodeTemplate
    metadata:
      name: default
      annotations:
        "helm.sh/hook": "post-install,post-upgrade"
        "helm.sh/hook-delete-policy": before-hook-creation
    spec:
      blockDeviceMappings:
        - deviceName: /dev/xvda
          ebs:
            volumeSize: 15Gi
            volumeType: gp3
            encrypted: false
      subnetSelector:
        aws-ids: "${worker_subnets}"
      securityGroupSelector:
        aws-ids: "${worker_security_groups}"
      tags:
        project: "coretools"
        environment: "prd"
        terraform: "true"
        Name: "karpenter-node-poc"
    

  - apiVersion: karpenter.sh/v1alpha5
    kind: Provisioner
    metadata:
      name: default
      annotations:
        "helm.sh/hook": "post-install,post-upgrade"
        "helm.sh/hook-delete-policy": before-hook-creation
    spec:
      # References cloud provider-specific custom resource, see your cloud provider specific documentation
      providerRef:
        name: default

      taints:
        - key: dedicated
          value: karpenter
          effect: NoSchedule

      labels:
        nodegroup: karpenter

      requirements:
        - key: karpenter.k8s.aws/instance-category
          operator: In
          values: [m, c]
        - key: "karpenter.k8s.aws/instance-hypervisor"
          operator: In
          values: ["nitro"]
        - key: karpenter.k8s.aws/instance-generation
          operator: Gt
          values: ["4"]
        - key: kubernetes.io/arch
          operator: In
          values: ["amd64"]
        - key: "karpenter.sh/capacity-type"
          operator: In
          values: ["on-demand"]

      # Enables consolidation which attempts to reduce cluster cost by both removing un-needed nodes and down-sizing those
      # that can't be removed.  Mutually exclusive with the ttlSecondsAfterEmpty parameter.
      consolidation:
        enabled: true

      # Resource limits constrain the total size of the cluster.
      # Limits prevent Karpenter from creating new instances once the limit is exceeded.
      limits:
        resources:
          cpu: "100"
          memory: 100Gi

      # Enables consolidation which attempts to reduce cluster cost by both removing un-needed nodes and down-sizing those
      # that can't be removed.  Mutually exclusive with the ttlSecondsAfterEmpty parameter.
      # consolidation:
      #   enabled: false

      # If omitted, the feature is disabled and nodes will never expire.  If set to less time than it requires for a node
      # to become ready, the node may expire before any pods successfully start.
      ttlSecondsUntilExpired: 2592000 # 30 Days = 60 * 60 * 24 * 30 Seconds;

      # If omitted, the feature is disabled, nodes will never scale down due to low utilization
      # ttlSecondsAfterEmpty: 30

      # Priority given to the provisioner when the scheduler considers which provisioner
      # to select. Higher weights indicate higher priority when comparing provisioners.
      # Specifying no weight is equivalent to specifying a weight of 0.
      # weight: 10
```

**Comment by bwagner5** [comment]
Ah ok, so this is related to consolidation rather than drift. I'll try to reproduce it, but I'd agree that Karpenter shouldn't result in disrupting all of your pods with the PDB you have setup.

**Comment by marlovil** [comment]
Hi @bwagner5 

Have you been able to reproduce it? If you need something else, please feel free to ask me!

Thanks!

**Comment by tzneal** [comment]
We have not been able to reproduce this.  Can you send us the output of:

`kubectl get provisioner default`

from before and after you modify it?

I'm wondering if your tooling is deleting the old provisioner and creating a new one which would cause all of the old nodes to be removed at once.

**Comment by marlovil** [comment]
Hi @tzneal 

Yes, you were right! Defining the Provisioner and AWSNodeTemplate inside the helm values (under `extraObjects` section) causes that both objects are recreated when a change is detected in the helm values.

I've seen that the `extraObjects` section in the values.yaml  and extra-manifestst.yaml template has been removed in the latest version (0.28.0-rc.1). 

I've defined the both objects (Provisioner and AWSNodeTemplate) in a separate Terraform kubernetes_manifest resource. Now, if something changes, the objects are no recreated and the pods are not deleted.

It would be nice if Karpenter could detect configuration drift in `AWSNodeTemplate` and `Provisioners` definitions and update all the nodes with new specs. I'll submit a feature request for that. :)

Thanks a lot for your help, @tzneal!

Marc
